mlir/lib/Dialect/Rock/Transforms/GemmOutputSwizzle.cpp

```diff
@@ -0,0 +1,571 @@
+  dimensionsToRemove.insert("mblock");
+  dimensionsToRemove.insert("nblock");
+  FailureOr<ArrayAttr> maybeIdToLDS =
+      removeUpperDims(b, srcTransform, dimensionsToRemove);
+  if (failed(maybeIdToLDS)) {
+    LLVM_DEBUG(llvm::dbgs() << "getIdToLDS failed\n");
+    return std::nullopt;
+  }
+  ArrayAttr idToLDS = maybeIdToLDS.value();
+
+  ArrayRef<int64_t> shape = getLowerShape(idToLDS);
+  if (shape.size() != 2) {
+    LLVM_DEBUG(llvm::dbgs() << "getIdToLDS failed\n");
+    return std::nullopt;
+  }
+  int64_t mPerBlock = shape[0];
+  int64_t nPerBlock = shape[1];
+
+  return std::make_tuple(mPerBlock, nPerBlock, idToLDS);
+}
+
+struct ThreadwiseWriteAllRewritePattern
+    : public OpRewritePattern<ThreadwiseWriteAllOp> {
+  using OpRewritePattern<ThreadwiseWriteAllOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(ThreadwiseWriteAllOp op,
+                                PatternRewriter &b) const override {
+    Location loc = op.getLoc();
+
+    // Prepare some useful constants.
+    Value convertedC = op.getSource();
+    Value matC = op.getDest();
+    Type destType = op.getDest().getType().getElementType();
+
+    // Convert from reg -> memory transform to reg -> block
+    int64_t mPerBlock, nPerBlock;
+    ArrayAttr idToLDS;
+    std::optional<std::tuple<int64_t, int64_t, ArrayAttr>> maybeBlockInfo =
+        getIdToLDS(op, b);
+    if (!maybeBlockInfo.has_value()) {
+      return failure();
+    }
+    std::tie(mPerBlock, nPerBlock, idToLDS) = maybeBlockInfo.value();
+
+    // Obtain critical matrix dimensions.
+    ArrayRef<int64_t> cShape;
+    cShape = op.getDest().getType().getShape();
+    if (cShape.size() != 3) {
+      return failure();
+    }
+    int64_t G = cShape[0];
+    int64_t M = cShape[1];
+    int64_t N = cShape[2];
+    LLVM_DEBUG(llvm::dbgs() << "G: " << G << " N: " << N << " M: " << M
+                            << " nPerBlock: " << nPerBlock
+                            << " mPerBlock: " << mPerBlock << "\n");
+    FailureOr<IntegerAttr> maybeGridSize = getGridSize(op);
+    if (failed(maybeGridSize)) {
+      return failure();
+    }
+    FailureOr<IntegerAttr> maybeBlockSize = getBlockSize(op);
+    if (failed(maybeBlockSize)) {
+      return failure();
+    }
+    int64_t blockSize = maybeBlockSize.value().getValue().getSExtValue();
+    int64_t gridSize = maybeGridSize.value().getValue().getSExtValue();
+    bool useIndexDiffs = true;
+    bool forceUnroll = true;
+    int64_t ldsRequiredBytes = mPerBlock * nPerBlock * getByteWidth(destType);
+
+    // Decide register vectorization.
+    constexpr int64_t dimensionM = 1;
+    constexpr int64_t dimensionN = 2;
+    int64_t dataPerThread = (nPerBlock * mPerBlock) / blockSize;
+    if ((nPerBlock * mPerBlock) % blockSize != 0) {
+      return failure();
+    }
+
+    VectorizationResult mVectorRes =
+        getMaxVectorization(matC, dimensionM, /*inputDimLen=*/
+                            dataPerThread, matC.getDefiningOp());
```

Comment:
You should not need to give dataPerThread here. Thus, all the dependent information extracted above should not be needed.

Reply:
but feel free to explain if I am wrong.

Reply:
ok after reading a bit more you need dataPerThread  for something else... but please extract properly from `threadwise_write_all`

Reply:
I think this is related to a previous comment, dataPerThread = (nPerBlock * mPerBlock) / blockSize. And nPerBlock, mPerBlock are extracted from threadwise_write_all. Is this correct?

Reply:
These are for testing whether the pass should be applied, and `dataPerThread` is "how muth data would be copied per thread if this pass were applied"